pkg/services/object/server.go

```diff
@@ -1177,27 +1212,31 @@ func continueGetFromRemoteNode(ctx context.Context, conn *grpc.ClientConn, req *
 			if err != nil {
 				return err
 			}
-			onceHdr.Do(func() {
-				err = stream.WriteHeader(obj)
+			x.onceHdr.Do(func() {
+				err = x.respStream.WriteHeader(obj)
 			})
 			if err != nil {
 				return fmt.Errorf("could not write object header in Get forwarder: %w", err)
 			}
+
+			x.payloadHashCheck = v.Init.Header.PayloadHash.Sum
+			x.payloadHashGot = sha256.New()
 		case *protoobject.GetResponse_Body_Chunk:
 			if !headWas {
 				return errors.New("incorrect message sequence")
 			}
 			fullChunk := v.Chunk
-			respChunk := chunkToSend(*respondedPayload, readPayload, fullChunk)
+			respChunk := chunkToSend(x.respondedPayload, readPayload, fullChunk)
 			if len(respChunk) == 0 {
 				readPayload += len(fullChunk)
 				continue
 			}
-			if err := stream.WriteChunk(respChunk); err != nil {
+			if err := x.respStream.WriteChunk(respChunk); err != nil {
 				return fmt.Errorf("could not write object chunk in Get forwarder: %w", err)
 			}
 			readPayload += len(fullChunk)
-			*respondedPayload += len(respChunk)
+			x.respondedPayload += len(respChunk)
+			x.payloadHashGot.Write(respChunk) // never returns an error according to docs
```

Comment:
Maybe it can be checked here? We always know the proper length.

Reply:
to me, it's pretty expected to verify checksum when the data is over. Normally, we'll continue and get EOF insta

we dont check for truncated/excessive payload here, regardless of hash. Seems like it's worth doing

Reply:
done